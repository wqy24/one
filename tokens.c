#include "tokens.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>
#include <stdint.h>

#define EXPAND_RATE 32

int pred_ref(char c) { return c != '}'; }
int pred_str(char c) { return c != '`'; }
int pred_num(char c) { return (isdigit(c) || c == '.'); }
int pred_idn(char c) { return (isalnum(c) || c == '_' || c == '-'); }

char *read_value(char first, FILE *input, int (*pred)(char), int rm_quote)
{
  size_t size = EXPAND_RATE * sizeof(char);
  char *buffer = (char *)malloc(size), *res;
  char c;
  size_t count = 0;
  if (!rm_quote) buffer[count++] = first;

  while ((c = getc(input)) != EOF && pred(c)) {
    buffer[count++] = c;
    if (count + 1 >= size) {
      size += EXPAND_RATE * sizeof(char);
      char *tmp = realloc(buffer, size);
      if (tmp == NULL) {
        fputs("Out of memory\n", stderr);
        free(buffer);
        return NULL;
      }
      buffer = tmp;
    }
  }

  if (rm_quote && pred(c)) {
    fputs("EOF while parsing a quoted token\n", stderr);
    free(buffer);
    return NULL;
  }

  buffer[count] = '\0';
  res = realloc(buffer, (count + 1) * sizeof(char));
  if (res == NULL) {
    fputs("Out of memory\n", stderr);
    free(buffer);
    return NULL;
  }

  if (!rm_quote && !pred(c)) {
    fseek(input, -1, SEEK_CUR);
  }

  return res;
}

one_token read_one_token(FILE *input)
{
  char curr;

  switch (curr = getc(input)) {
    case EOF: return (one_token){FEOF, NULL};
    case '@': return (one_token){AT, NULL};
    case '^': return (one_token){CARET, NULL};
    case '&': return (one_token){ET, NULL};
    case '(': return (one_token){L_PRN, NULL};
    case ')': return (one_token){R_PRN, NULL};
    case '[': return (one_token){L_BRK, NULL};
    case ']': return (one_token){R_BRK, NULL};
    case '|': return (one_token){PIPE, NULL};
    case ':': return (one_token){COLON, NULL};
    case '<': return (one_token){LT, NULL};
    case '>': return (one_token){GT, NULL};
    case '?': return (one_token){QUESTION, NULL};
    case '~': return (one_token){WAVE, NULL};
    case ';':
      while(!feof(input) && getc(input) != '\n') /* do nothing */;
      return read_one_token(input);
    case ',': return (one_token){COMMA, NULL};
    case '{': return (one_token){REF, read_value('{', input, pred_ref, 1)};
    case '`': return (one_token){STR, read_value('`', input, pred_str, 1)};
    default:
      if (isspace(curr)) return read_one_token(input);
      if (isdigit(curr) || curr == '-') return (one_token){NUM, read_value(curr, input, pred_num, 0)};
      if (isalpha(curr) || curr == '_') return (one_token){IDN, read_value(curr, input, pred_idn, 0)};
      fprintf(stderr, "Invalid character %c.\n", curr);
      exit(EXIT_FAILURE);
  }
}
