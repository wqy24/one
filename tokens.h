#ifndef __ONE_TOKENS_H__
#define __ONE_TOKENS_H__

#include <stdio.h>

typedef struct token {
  enum {
    FEOF,
    STR,
    NUM,
    REF,
    IDN,
    L_BRK,
    R_BRK,
    L_PRN,
    R_PRN,
    PIPE,
    CARET,
    COLON,
    COMMA,
    LT, GT,
    ET /* & */, // AT /* @ */,
    QUESTION,
    WAVE
  } type;
  char *value;
} one_token;

one_token read_one_token(FILE *input); /* malloc warning: "value" in the token */

#endif
