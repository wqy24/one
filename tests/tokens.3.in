12.3
`Hold it!` 
; Objection!`
param^num: add 1 param;@
abs -124 -124| abs [2,4,8]
?a: elem^a: [elem]
(abs -990)
