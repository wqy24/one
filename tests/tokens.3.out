NUM
12.3
STR
Hold it!
IDN
param
CARET
IDN
num
COLON
IDN
add
NUM
1
IDN
param
IDN
abs
NUM
-124
NUM
-124
PIPE
IDN
abs
L_BRK
NUM
2
COMMA
NUM
4
COMMA
NUM
8
R_BRK
QUESTION
IDN
a
COLON
IDN
elem
CARET
IDN
a
COLON
L_BRK
IDN
elem
R_BRK
L_PRN
IDN
abs
NUM
-990
R_PRN
FEOF
